engine/flink/management/src/main/scala/pl/touk/nussknacker/engine/management/FlinkDeploymentManager.scala

```diff
@@ -14,16 +14,31 @@ import pl.touk.nussknacker.engine.canonicalgraph.CanonicalProcess
 import pl.touk.nussknacker.engine.deployment.{DeploymentData, ExternalDeploymentId, User}
 import pl.touk.nussknacker.engine.management.FlinkDeploymentManager.prepareProgramArgs
 import ModelData._
+import pl.touk.nussknacker.engine.api.deployment.inconsistency.InconsistentStateDetector
 
 import scala.concurrent.{ExecutionContext, Future}
 
-abstract class FlinkDeploymentManager(modelData: BaseModelData, shouldVerifyBeforeDeploy: Boolean, mainClassName: String)(implicit ec: ExecutionContext)
+abstract class FlinkDeploymentManager(modelData: BaseModelData, shouldVerifyBeforeDeploy: Boolean, mainClassName: String)
+                                     (implicit ec: ExecutionContext, deploymentService: ProcessingTypeDeploymentService)
   extends DeploymentManager with AlwaysFreshProcessState with LazyLogging {
 
   private lazy val testRunner = new FlinkProcessTestRunner(modelData.asInvokableModelData)
 
   private lazy val verification = new FlinkProcessVerifier(modelData.asInvokableModelData)
 
+  /**
+    * Gets status from engine, handles finished state, resolves possible inconsistency with lastAction and formats status using `ProcessStateDefinitionManager`
+    */
+  override def getProcessState(name: ProcessName, lastAction: Option[ProcessAction])(implicit freshnessPolicy: DataFreshnessPolicy): Future[WithDataFreshnessStatus[Option[ProcessState]]] = {
+    val stefan: Future[WithDataFreshnessStatus[Option[ProcessState]]] = for {
+      statusWithFreshness <- getProcessState(name)
+      finishedStatusOpt = statusWithFreshness.value.filter(_.status.isFinished)
+      cancelActionOpt <- finishedStatusOpt.map(_ => deploymentService.markProcessFinishedIfLastActionDeploy(name)).sequence.map(_.flatten)
```

Comment:
Let's move this comment from DeploymentService here:
```
  // There is small problem here: if no one invokes process status for long time, Flink can remove process from history
  //- then it's gone, not finished.
```
And maybe add TODO that it should be checked periodically instead of for each `getProcessState` invocation

Reply:
Done. 

Regarding TODO: 

We should consider combining `InconsistentStateDetector` with `markProcessFinishedIfLastActionDeploy` (...aaand with `waitForDuringDeployFinished`) to have one place in `DeploymentManager` where "status-from-engine" is compared with and adjusted with "status-from-designer".

I've mentioned `waitForDuringDeployFinished` because of confusing error that occured in my real-life deployment: "Deploy execution finished, but job is still in during deploy state on Flink". Here we are checking periodically, for a limited time, eventually failed deployment leads to this error.

Reply:
Brilliant idea, I hope we find time someday to make it happen.